### PR TITLE
Add ATen export and gradient for torch.max/min

### PR DESCRIPTION
**Pull request by mindest**
**Description**: Add ATen export and gradient for `torch.max`/`min`

**Motivation and Context**
There are currently three cases in exporting `torch.max`:
1. `torch.max(input)`, exported as `ReduceMax`
2. `torch.max(input, other)`, exported as `Max`
3. `torch.max(input, dim, keepdim)`, exported as `ReduceMax`+`ArgMax`

We already have gradient for `Max`. This change deals with cases 1 and 3. Likewise for `torch.min`.

### CODE REVIEW

File: onnxruntime/python/tools/symbolic_shape_infer.py [hunk 1]
```diff
@@ -1151,6 +1153,28 @@ def _infer_aten_pool2d(self, node):
             elem_type = onnx.TensorProto.INT64 if i == 1 else self.known_vi_[node.input[0]].type.tensor_type.elem_type
             vi.CopyFrom(helper.make_tensor_value_info(o, elem_type, get_shape_from_sympy_shape(sympy_shape)))
 
+    def _infer_aten_minmax(self, node):
+        dim_or_y = self._try_get_value(node, 1)
+        keepdim = self._try_get_value(node, 2)
```
Comment (iK1D) [inline onnxruntime/python/tools/symbolic_shape_infer.py:1158]: None doesn't mean input[1] and input[2] don't exist, it can be some non-constant inputs. I think use len(node.input) is enough to switch the cases, if 1 is max, if 3, then max.dim. If len(node.input) is 3 and dim or keepdims is None, it means unkown axis, it's still max.dim, there are some code in _infer_ReduceSum you can borrow to handle such case.

File: orttraining/orttraining/python/training/ortmodule/_custom_gradient_registry.py [hunk 2]
```diff
@@ -109,6 +109,34 @@ def max_pool2d_gradient():
...
 
 
+def minmax_gradient():
+    # Gradient of torch.min(input) (and max)
+    return [
+        ('Equal', ['I(0)', 'O(0)'], ['Mask']),
+        ('Constant', [], ['Const_0'], {'value': {'value': 0., 'dtype': 'IElemType(0)', 'is_tensor': True}}),
+        ('Constant', [], ['Const_1'], {'value': {'value': 1., 'dtype': 'IElemType(0)', 'is_tensor': True}}),
+        ('Where', ['Mask', 'Const_1', 'Const_0'], ['MaskValue']),
+        ('ReduceSum', ['MaskValue'], ['MaskSum']),
```
Comment (iK1D) [inline orttraining/orttraining/python/training/ortmodule/_custom_gradient_registry.py:119]: ONNX's ReduceSum's keepdim is true by default, if you don't set the keepdim attr to false here, the result will not be scalar, which might have perf impact to following calculation as it might need broadcasting offset calculation. Not sure if the binary elementwise already handle all-1 dims or not, better add the keepdim attr to false here.

File: orttraining/orttraining/python/training/ortmodule/_custom_gradient_registry.py [hunk 3]
```diff
@@ -109,6 +109,34 @@ def max_pool2d_gradient():
     ]
 
 
+def minmax_gradient():
```
Comment (iK1D) [inline orttraining/orttraining/python/training/ortmodule/_custom_gradient_registry.py:112]: Better add comment to explain PyTorch's gradient logics are different for these two cases, so we also separate handling these cases.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 4]
```diff
@@ -1003,6 +1003,39 @@ def run_step(model, input):
 
         _test_helpers.assert_values_are_close(ort_prediction, pt_prediction)
 
+@pytest.mark.parametrize("operator", ['min', 'max'])
+@pytest.mark.parametrize("dim", [None, 0, -1])
+@pytest.mark.parametrize("keepdim", [True, False])
```
Comment (iK1D) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:1008]: Better also test keepdim is None (i.e., Max(x,y)) case, maybe add a new test case for it. As you also move such case from PyTorch exporter to your above code.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 5]
```diff
@@ -1003,6 +1003,39 @@ def run_step(model, input):
...
+def test_gradient_correctness_max(operator, dim, keepdim):
+    func = torch.min if operator == 'min' else torch.max
+
+    class NeuralNetMax(torch.nn.Module):
+        def forward(self, input):
+            if dim is None:
+                return func(input)
```
Comment (iK1D) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:1015]: I am thinking adding one nore func to consume the outputs inside toe module is better. As max and max.dim are returning different number of outputs.

File: onnxruntime/python/tools/symbolic_shape_infer.py [hunk 6]
```diff
@@ -1151,6 +1153,30 @@ def _infer_aten_pool2d(self, node):
             elem_type = onnx.TensorProto.INT64 if i == 1 else self.known_vi_[node.input[0]].type.tensor_type.elem_type
             vi.CopyFrom(helper.make_tensor_value_info(o, elem_type, get_shape_from_sympy_shape(sympy_shape)))
 
+    def _infer_aten_minmax(self, node):
+        vi = self.known_vi_[node.output[0]]
+        if len(node.input) == 1:
+            vi.CopyFrom(helper.make_tensor_value_info(
+                node.output[0], self.known_vi_[node.input[0]].type.tensor_type.elem_type, []))
+        else: # len(node.input) == 3
```
Comment (pengwa) [inline onnxruntime/python/tools/symbolic_shape_infer.py:1161]: add an assert for len == 3?

**Review by pengwa — COMMENTED** [review]
LGTM, as talked offline, would be nice if we have one more test case covering additional branch of infer shape.

File: orttraining/orttraining/python/training/ortmodule/_custom_gradient_registry.py [hunk 7]
```diff
@@ -109,6 +109,45 @@ def max_pool2d_gradient():
...
+    # In PyTorch, when there are multiple maxima/minima along the axis, the gradient is granted to the first.
+    # e.g.: x = torch.tensor([1., 2., 2., 1., 2.], requires_grad=True)
+    #       y = x.max(dim=0)     (torch.return_types.max(values=tensor(2., grad_fn=<MaxBackward0>),
+    #                                                    indices=tensor(1)))
+    #       y.values.backward()
+    #       print(x.grad)        (tensor([0., 1., 0., 0., 0.]))
+    return [
+        ('Shape', ['I(0)'], ['Shape_X']),
+        (('ATen', 'org.pytorch.aten'), ['GO(0)', 'I(1)', 'O(1)', 'Shape_X', 'I(2)'], ['GI(0)'],
+         {'operator': {'value': 'aten::value_selecting_reduction_backward', 'dtype': 'string'}})
```
[the remark this answers is not in the document]
Comment (iK1D) [reply orttraining/orttraining/python/training/ortmodule/_custom_gradient_registry.py:144]: because the attribute 'operator' is string type attribute, and the value is 'aten::value_selecting_reduction_backward'

**Review by pengwa — APPROVED** [review]

**Review by iK1D — APPROVED** [review]

**Comment by mindest** [comment]
Temporarily removed the UT for `torch.max(input_tensor, other_tensor)` case because torch's gradient implementation for this case has been changed (for the two maxima case, tested on torch1.9.0 and torch1.11.0), which causes UT to fail where the two tensors have the same value. Will create another PR to modify the `MaxGradient` and keep consistent with torch. Details will be given there.

**Comment by pengwa** [comment]
> Temporarily removed the UT for `torch.max(input_tensor, other_tensor)` case because torch's gradient implementation for this case has been changed (for the two maxima case, tested on torch1.9.0 and torch1.11.0), which causes UT to fail where the two tensors have the same value. Will create another PR to modify the `MaxGradient` and keep consistent with torch. Details will be given there.

Interesting, do you think why they do the change?

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 8]
```diff
@@ -1123,37 +1123,6 @@ def run_step(model, input):
         _test_helpers.assert_values_are_close(ort_input.grad, pt_input.grad)
 
 
-@pytest.mark.parametrize("operator", ["min", "max"])
```
Comment (pengwa) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py]: I think you can use some feature of pytest  to skip this test currently with comments.

**Comment by mindest** [comment]
> > Temporarily removed the UT for `torch.max(input_tensor, other_tensor)` case because torch's gradient implementation for this case has been changed (for the two maxima case, tested on torch1.9.0 and torch1.11.0), which causes UT to fail where the two tensors have the same value. Will create another PR to modify the `MaxGradient` and keep consistent with torch. Details will be given there.
> 
> Interesting, do you know why they do the change?

Related discussions: https://github.com/pytorch/pytorch/issues/56734, https://github.com/pytorch/pytorch/pull/59669.
But no reason was given why the old implementation of gradient was wrong.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py (hunk 8 above)
Comment (mindest) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py]: Thanks for the reminder! To disable for now is better.

**Review by pengwa — DISMISSED** [review]

**Review by pengwa — APPROVED** [review]
